### PR TITLE
Feature/ms 10 download s3

### PR DESCRIPTION
**Pull request by rhoxhaj**
## Abstract
Adapt endpoints used to get binary data:

    https://getmesh.io/docs/api/#project__webrootfield__fieldName___path__get
    https://getmesh.io/docs/api/#project__webroot__path__get
    https://getmesh.io/docs/api/#project__nodes__nodeUuid__binary__fieldName__get

In each case, the response should be a redirect 302 with the presigned URL. There should be possible also to manipulate images when requested via the parameters.

### CODE REVIEW

File: core/src/main/java/com/gentics/mesh/core/data/container/impl/NodeGraphFieldContainerImpl.java [hunk 1]
```diff
@@ -724,10 +722,15 @@ public String getSegmentFieldValue() {
...
 		if (stringField == null) {
-			BinaryGraphField binaryField = getBinary(segmentFieldKey);
-			if (binaryField != null) {
-				return binaryField.getFileName();
+			S3BinaryGraphField s3binaryField = getS3Binary(segmentFieldKey);
+			if (nonNull(s3binaryField) && nonNull(s3binaryField.getS3Binary())) {
+				return s3binaryField.getS3Binary().getFileName();
 			}
+			BinaryGraphField binary = getBinary(segmentFieldKey);
+			if (nonNull(binary) && nonNull(binary.getBinary())) {
```
Comment (plyhun) [inline core/src/main/java/com/gentics/mesh/core/data/container/impl/NodeGraphFieldContainerImpl.java:730]: Are you sure regarding the extra check for `binary.getBinary()`?

File: core/src/main/java/com/gentics/mesh/core/data/node/impl/NodeImpl.java [hunk 2]
```diff
@@ -1985,6 +1986,22 @@ private PathSegment getSegment(String branchUuid, ContainerType type, String seg
 					return new PathSegmentImpl(container, binaryField, container.getLanguageTag(), segment);
 				}
 			}
+			//TODO S3
```
Comment (plyhun) [inline core/src/main/java/com/gentics/mesh/core/data/node/impl/NodeImpl.java:1989]: Is there something left here to do?

File: core/src/main/java/com/gentics/mesh/core/endpoint/node/S3UploadContext.java [hunk 3]
```diff
@@ -30,6 +30,7 @@ public void setS3ObjectKey(String s3ObjectKey) {
 		this.s3ObjectKey = s3ObjectKey;
 	}
```
Comment (plyhun) [inline core/src/main/java/com/gentics/mesh/core/endpoint/node/S3UploadContext.java:32]: This line change may be reverted.

File: core/src/main/java/com/gentics/mesh/core/endpoint/webroot/WebRootHandler.java [hunk 4]
```diff
@@ -98,7 +103,13 @@ public void handleGetPath(RoutingContext rc) {
 				}
 				binaryFieldResponseHandler.handle(rc, binaryField);
 				return null;
```
Comment (plyhun) [inline core/src/main/java/com/gentics/mesh/core/endpoint/webroot/WebRootHandler.java:105]: Here may be `else if`.

File: mdm/common/src/main/java/com/gentics/mesh/core/image/ImageManipulator.java [hunk 5]
```diff
@@ -1,60 +1,70 @@
...
+import java.awt.image.BufferedImage;
+import java.io.InputStream;
+import java.util.Map;
+
 /**
  * SPI provider interface for image manipulators.
  */
 public interface ImageManipulator {
 
-	/**
```
Comment (plyhun) [inline mdm/common/src/main/java/com/gentics/mesh/core/image/ImageManipulator.java:17]: Something happened to this file's indentation - it looks like unchanged code-wise.

**Comment by plyhun** [comment]
Except abovementioned, all looks ok. Before merging the whole S3 into the target release branch I will anyway be checking everything for working and passing the unit tests.

File: core/src/main/java/com/gentics/mesh/core/data/container/impl/NodeGraphFieldContainerImpl.java (hunk 1 above)
Comment (rhoxhaj) [reply core/src/main/java/com/gentics/mesh/core/data/container/impl/NodeGraphFieldContainerImpl.java:730]: Actually I found this relevant for S3BinaryField but if binary.getBinary() has worked for binary fields, it is better to revert this change. I will revert it and check only if this is not null.

File: core/src/main/java/com/gentics/mesh/core/data/node/impl/NodeImpl.java (hunk 2 above)
Comment (rhoxhaj) [reply core/src/main/java/com/gentics/mesh/core/data/node/impl/NodeImpl.java:1989]: No, I had forgot to remove this. Done now :)

**Pull request merged**